### PR TITLE
docs: add a coding challenge to build a blogging platform through a REST APIs

### PR DESCRIPTION
**Pull request by berat-552**
Added a new coding challenge to the README.

It's a challenge which includes an API with 2 models.

JSON Web Tokens and password hashing is used to provide security.

### CODE REVIEW

File: coding-challenges/README.md [hunk 1]
```diff
@@ -47,3 +47,69 @@ Sorting algorithm should correctly sort the products in ascending order based on
...
+**Context**
+
+AlphaTech is developing a new microservice for a blogging platform. They want to implement a secure authentication mechanism using JWT to ensure only authenticated users can create, edit, and delete blog posts. User password must be hashed to provide security.
+
+**Requirements**
+
+JWT Authentication: 
+
+- Implement endpoints for user sign up, login, logout and delete.
+- Upon successful authentication (login), issue a JWT containing user information and necessary claims (e.g. user ID, iat, exp)
```
Comment (berat-552) [inline coding-challenges/README.md:63]: just realised, it should be `(e.g. user ID, accessToken, iat, exp)`

File: coding-challenges/README.md [hunk 2]
```diff
@@ -47,3 +47,69 @@ Sorting algorithm should correctly sort the products in ascending order based on
...
+Protected routes:
+- Create a blog
+- Read a blog
+- Update a blog
+- Delete a blog
+- List all blogs of user
+
+Protected routes:
+
+Data Structures: use simple struct for a user, which includes:
```
Comment (berat-552) [inline coding-challenges/README.md:81]: should be `use simple struct for a user and blog`

**Comment by AbuHafsa** [comment]
@moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)

**Comment by moabukar** [comment]
> @moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)
> 
>  

@AbuHafsa Yeah I see what you mean. However, I think we should keep this kind of challenge vague enough for candidates to figure out which library they should use. You can hint towards some libraries but keep it open and language agnostic. 

This way they can research and find out whether they use gorilla, gin etc or any other capable framework. 

What do you think?

**Comment by berat-552** [comment]
> > @moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)
> 
> @AbuHafsa Yeah I see what you mean. However, I think we should keep this kind of challenge vague enough for candidates to figure out which library they should use. You can hint towards some libraries but keep it open and language agnostic.
> 
> This way they can research and find out whether they use gorilla, gin etc or any other capable framework.
> 
> What do you think?
> 
> Cc @berat-552

@moabukar I also agree, while i was writing up this challenge, i removed some information with the intent that the person taking the challenge to figure out some things by themselves.

**Comment by moabukar** [comment]
> > > @moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)
> 
> > 
> 
> > @AbuHafsa Yeah I see what you mean. However, I think we should keep this kind of challenge vague enough for candidates to figure out which library they should use. You can hint towards some libraries but keep it open and language agnostic.
> 
> > 
> 
> > This way they can research and find out whether they use gorilla, gin etc or any other capable framework.
> 
> > 
> 
> > What do you think?
> 
> > 
> 
> > Cc @berat-552
> 
> 
> 
> @moabukar I also agree, while i was writing up this challenge, i removed some information with the intent that the person taking the challenge to figure out some things by themselves.

Feel free to adjust the constraints. Right now it's limited to Golang 😅

**Comment by AbuHafsa** [comment]
> > @moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)
> 
> @AbuHafsa Yeah I see what you mean. However, I think we should keep this kind of challenge vague enough for candidates to figure out which library they should use. You can hint towards some libraries but keep it open and language agnostic.
> 
> This way they can research and find out whether they use gorilla, gin etc or any other capable framework.
> 
> What do you think?
> 
> Cc @berat-552

Yeah I think that's fair to give some hints but let them pick any of the available out there. I also think it is a good choice to go with Go because you get a taste of low-level language at a comfort of a high-level one with type inference and other features.

**Comment by berat-552** [comment]
I have added a couple of words expressing that they can use something like spring boot or quarkus, hope thats fine

**Comment by moabukar** [comment]
> > > @moabukar the brother used the first challenge as a template but I see that his proposed challenge is a bit tricky to build with only Gorilla Mux, should he go with a JVM-based framework instead? (i.e Quarkus, Spring Boot, ..)
> 
> > 
> 
> > @AbuHafsa Yeah I see what you mean. However, I think we should keep this kind of challenge vague enough for candidates to figure out which library they should use. You can hint towards some libraries but keep it open and language agnostic.
> 
> > 
> 
> > This way they can research and find out whether they use gorilla, gin etc or any other capable framework.
> 
> > 
> 
> > What do you think?
> 
> > 
> 
> > Cc @berat-552
> 
> 
> 
> Yeah I think that's fair to give some hints but let them pick any of the available out there. I also think it is a good choice to go with Go because you get a taste of low-level language at a comfort of a high-level one with type inference and other features.

Yeah sounds good. For sure, Go is definitely a good choice but we don't want to restrict the candidate to just Go. There might be some learning Python or JS for example and they can use flask or express etc. 

Perhaps moving forward with the challenges, we can have a recommended section but keep it open and vague enough for people to solve it how they want. Just have an overall acceptance criteria etc

File: coding-challenges/README.md [hunk 3]
```diff
@@ -47,3 +47,69 @@ Sorting algorithm should correctly sort the products in ascending order based on
...
+
+Bonus: 
+- Implement Rate Limiting middleware.
+- Implement a Regular Expression pattern for password validation.
+
+Documentation: Provide a README that explains how to run your code and use your API and include instructions on how to interact with the authentication endpoints.
+
+**Constraints**
+
+Stick to Golang's standard library if possible. You can use a package like Gorilla Mux for routing, or a desired backend framework of your choice (e.g. Quarkus, Spring Boot, Flask, ExpressJS). Keep your application as stateless as possible. Please make sure to comment your code.
```
Comment (moabukar) [inline coding-challenges/README.md:109]: You sure you want to keep this line? Why stick to Golang? Let peeps solve it in Go or JS or Python etc

Comment (berat-552) [reply coding-challenges/README.md:109]: i'll remove the constraints and that the golang part

File: coding-challenges/README.md [hunk 4]
```diff
@@ -46,7 +46,7 @@ All endpoints should work without any errors.
 Sorting algorithm should correctly sort the products in ascending order based on price.
 Test cases should cover basic scenarios and edge cases for the sorting algorithm.
 
-```
+```bash
```
Comment (berat-552) [inline coding-challenges/README.md:49]: added bash on the code block because coding-challenge-1 had it, to keep things consistent

Comment (AbuHafsa) [reply coding-challenges/README.md:49]: If you look at the formatting of the README, it is somewhat messed up. You can see what it looks like under this link
https://github.com/berat-552/tech-vault/tree/feature-coding-challenge-2/coding-challenges

**Review by AbuHafsa — CHANGES_REQUESTED** [review]

**Comment by berat-552** [comment]
@AbuHafsa fixed it

**Comment by AbuHafsa** [comment]
Looking good alhamdulillah

**Review by AbuHafsa — APPROVED** [review]
